contributors/design-proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,302 @@
+
+---
+
+Verb: GET
+Namespace: webapp
+APIGroup: custom-metrics
+APIVersion: v1alpha1
+Resource: ingress.extensions
+Subresource: hits-per-second
+Name: ResourceAll(*)
+```
+
+Notice that getting metrics which describe a namespace follows a slightly
+different pattern from other resources; Since namespaces cannot feasibly
+have unbounded subresource names (due to collision with resource names,
+etc), we introduce a pseudo-resource named "metrics", which represents
+metrics describing namespaces, where the resource name is the metric name:
+
+```
+GET /apis/custom-metrics/v1alpha1/namespaces/webapp/metrics/queue-length
+
+---
+
+Verb: GET
+Namespace: webapp
+APIGroup: custom-metrics
+APIVersion: v1alpha1
+Resource: metrics
+Name: queue-length
+```
+
+API Path Design, Discovery, and Authorization
+---------------------------------------------
+
+The API paths in this proposal are designed to a) resemble normal
+kubernetes APIs, b) facilitate writing authorization rules, and c)
+allowing for discovery.
+
+Since the API structure follows the same structure as other Kubernetes
+APIs, it allows for fine grained control over access to metrics.  Access
+can be controlled on a per-metric basic (each metric is a subresource, so
+metrics may be whitelisted by allowing access to a particular
+resource-subresource pair), or granted in general for a namespace (by
+allowing access to any resource in the `custom-metrics` API group). 
+
+Similarly, since metrics are simply subresources, a normal Kubernetes API
+discovery document can be published by the adapter's API server, allowing
+clients to discover the available metrics.
+
+Note that we introduce the syntax of having a name of ` * ` here since
+there is no current syntax for getting the output of a subresource on
+multiple objects.
+
+API Objects
+-----------
+
+The request URLs listed above will return either the `MetricValueList` or
+`MetricValue` types (depending on whether they describe multiple objects,
+or a single one), described below:
+
+```go
+
+// a list of values for a given metric for some set of objects
+type MetricValueList struct {
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the value of the metric across the described objects
+    Items []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type MetricValue struct {
+    // a reference to the described object
+    DescribedObject ObjectReference `json:"describedObject"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // indicates the time at which the metrics were produced
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the window ([Timestamp-Window, Timestamp]) from
+    // which these metrics were calculated, when returning rate
+    // metrics calculated from cumulative metrics (or zero for
+    // non-calculated instantaneous metrics).
+    Window  unversioned.Duration `json:"window"`
```

Comment:
From past discussion my take away was that the exact semantic meaning is left to the provider of the custom metrics. Integrating systems behave differently and there's little point in specifying something here we can't enforce anyway.

I'd consider `window` as a return attribute a gimmick. It's much more likely that one wants to specify it in the request as the aggregation window has significant effect on the value.
Let's say we implemented a proxy for Prometheus. We'd follow the recording rules (materialized queries) naming pattern like:

`/...//namespaces/webapp/ingress.extensions/*/requests:rate5m?labelSelector=app%3Dfrontend`

The window is specified in the requested metric and the proxy would merely copy that into the `window` field in the response.

How is `window` used in the HPA? It seems tedious to extract from the metric name in this example – especially if users violate naming conventions.